### PR TITLE
Add IsStoppedBTNode

### PR DESCRIPTION
**Pull request by tonynajjar**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (add tickets here #1) |
| Primary OS tested on | (Ubuntu, MacOS, Windows) |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |
| Does this PR contain AI generated software? | (No; Yes and it is marked inline in the code) |

---

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [x] Check that any new parameters added are updated in docs.nav2.org
- [x] Check that any significant change is added to the migration guide
- [x] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [x] Check that any new functions have Doxygen added
- [x] Check that any new features have test coverage
- [x] Check that any new plugins is added to the plugins page
- [x] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by tonynajjar** [comment]
@SteveMacenski are you open to this contribution? Still a draft but almost there

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_stopped_condition.hpp
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_stopped_condition.hpp:97]: `OdomSmoother` nav2 util instead? I believe this is handed to the navigators as well on the BT blackboard so you don't need duplicate subscriptions. You can use the one already at play https://github.com/ros-navigation/navigation2/blob/main/nav2_bt_navigator/src/bt_navigator.cpp#L126-L129

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Groot (https://github.com/ros-navigation/navigation2/blob/main/nav2_behavior_tree/nav2_tree_nodes.xml) and doc updates are needed (the usual bits!).

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_stopped_condition.hpp
Comment (tonynajjar) [reply nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_stopped_condition.hpp:97]: Hmm, for such a BT node I don't know if I want my odom to be smoothed though. If the velocity is 0, I want to know it as soon as possible. As for fluctuations, there is the `time_stopped_threshold` that accounts for that.

Side-note: I got inspired from IsStuckCondition which also doesn't use the OdomSmoother, is that an oversight?

File: nav2_util/include/nav2_util/odometry_utils.hpp
Comment (tonynajjar) [inline nav2_util/include/nav2_util/odometry_utils.hpp:73]: that's pretty important right? `vel_smooth_` is not atomically updated

File: nav2_behavior_tree/plugins/condition/is_stopped_condition.cpp
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/condition/is_stopped_condition.cpp:18]: This file is missing BT unit testing

File: nav2_util/include/nav2_util/odometry_utils.hpp
Comment (SteveMacenski) [reply nav2_util/include/nav2_util/odometry_utils.hpp:73]: Yeah - this is pretty old, back in the Intel + Steve days, so I think there was some R&D thinking and not production thinking back then

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
On terminal conditions (success, fail) shouldn't we update `stopped_stamp_`?

**Comment by tonynajjar** [comment]
> On terminal conditions (success, fail) shouldn't we update stopped_stamp_?

With "update" I assume you mean resetting to `rclcpp::Time(0, 0, RCL_ROS_TIME)`? For FAILURE  we do. For doing it at SUCCESS, I think it's a tradeoff:

- reset at SUCCESS:  Not stateful: Next time it's ticked, it will wait again for `duration_stopped`, which could be undesired or unexpected behavior. But maybe stateful nodes is an anti-pattern so this behavior is actually expected? 
- no reset at SUCCESS: Stateful: Next time it's ticked, it won't wait "since it's been stopped for long enough already". However there could be this bug:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1. BT node returns SUCCESS after it was stopped for long enough
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2. robot moves and stops again before the BT node is ticked again
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3. BT node returns SUCCESS because the moving in between was not considered

All things considered, I think you're right that resetting at success would be the better option

**Comment by SteveMacenski** [comment]
Agreed, I was thinking something too like `node_->get_clock()->now() - stopped_stamp_ > rclcpp::Duration(duration_stopped_)` will always be true if we left the branch of the BT that contains this node and then come back to it after some time, if we don't reset the time after terminal conditions. Then, its a single-shot of the velocity without using your duration piece.

With that said, I think `it will wait again for duration_stopped, which could be undesired or unexpected behavior` can be largely mitigated if you have your waiting duration to be sub-100ms -- about the perception of a person to change. 100ms though should give you some readings to work based off of so any single erroneous measurement doesn't get you. 

---

Also see https://github.com/ros-navigation/docs.nav2.org/pull/612#issuecomment-2504507376 (I know looking at comments after a PR is merged gets lost in the general "merge" notification, for me at least :wink: ).

**Review by SteveMacenski — APPROVED** [review]
LGTM - let me know if there's anything else here from that conversation of its good to go

Though, I see a couple of lines that seem easy to cover in unit testing that are not covered: 
- `twist.header.stamp = node_->get_clock()->now();`
- Each of the `!received_odom_` condition's internals

https://app.codecov.io/gh/ros-navigation/navigation2/pull/4764?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=ros-navigation

**Comment by tonynajjar** [comment]
@SteveMacenski it's theoretically ready from my side but I'll stress test it even more in the coming week so we can wait a bit before merging

**Comment by SteveMacenski** [comment]
Ok, please do! Let me know when you're ready

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**
